### PR TITLE
Msspoly dynamics

### PR DESCRIPTION
**Pull request by aespielberg**
This is code I have added which tries to make the dynamics work with msspoly. In this, I have disabled the mex step of rigidbodymanipulator compilation, in order to make the RBM compile. The general test is to do something like:

p = RigidBodyManipulator('some_parameterized_urdf.urdf');
params = p.getParamFrame().getPoly;
p = p.setParams(params);
[xdot, dxdot] = p.dynamics(some args); <--this will fail near the end when it has to take an inverse if set as msspoly.

In most cases I have put tags saying %ANDY change where I have changed what I think might be critical pieces of code; of course, this should be noticeable by a diff as well.

I have also created a few new utility functions which I thought would be useful in dealing with msspoly. They all have the suffix msspoly.m, and are under utils.

I am wondering if the correct thing here is not to provide msspoly support for this use case, but rather matlab symbolic.

Finally two common things I noticed:

-If people want msspoly to be supported, numel is a bad function to use most places in the code and should be removed. numel(a) can be replaced with length(a(:)) to much the same effect. The reason to not use numel is that with msspoly, it returns the number of variables, NOT the number of entries in the vector/matrix/tensor/whatever.

-Explicitly creating a matrix out of zeros() or something is a bad idea as well, because setting a submatrix fails with incompatible types (setting a part of a matrix of type double to a type msspoly). I've usually just defaulted to msspoly here since that seems to be more general, but perhaps that was a dumb thing to do. In some places, I've also avoided pre-allocating the matrix, which might not be a good idea with big manipulators.

### CODE REVIEW

File: .cproject [hunk 1]
```diff
@@ -0,0 +1,373 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (tkoolen) [inline .cproject]: please don't add your Eclipse settings.

File: .gitmodules [hunk 2]
```diff
@@ -2,3 +2,71 @@
 	path = cmake
 	url = https://github.com/RobotLocomotion/cmake.git
 	branch = master
+[submodule "spotless"]
```
Comment (tkoolen) [inline .gitmodules]: Did you mean to change .gitmodules?

File: .project [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tkoolen) [inline .project]: please don't add your Eclipse settings.

**Comment by tkoolen** [comment]
Could you clean this pull request up a bit? I think you have a lot of unintended changes, such as changes to submodules, Eclipse project files, and the test_aespielberg folder. Could you also remove the `%ANDY CHANGE` / `%ANDYCHANGE` comments? Thanks.

**Comment by tkoolen** [comment]
Ah, I think I understand what happened with the Eclipse project files and subproject commits. I cleaned up the history a bit for you by rebasing your commits onto master, thus getting rid of all of the nasty merge commits. See https://github.com/tkoolen/drake/tree/tk-rebase-aespielberg-msspoly_dynamics.

You'll probably want to check out tk-rebase-aespielberg-msspoly_dynamics, make sure that everything you worked on is still there and working, and then push it to your fork of drake. You can either push it to a new branch and open a new pull request against that branch, or if you're sure that everything's in order, force push to aespielberg/msspoly_dynamics and we can keep using this pull request.

Could you still get rid of the test_aespielberg folder and remove the ANDY CHANGE comments?

**Comment by aespielberg** [comment]
Hi Twan,

I will clean up the comments and get rid of the ANDY CHANGE and repush
later otnight.  I want to be clear that I still don't have the mex problem
solved - like I've mentioned, I have a guess as to the cause of the problem
but I was hoping someone more knowledgeable with that part of the codebase
could verify.

I'm really sorry about all the Eclipse files and subproject commits - but
thanks for cleaning it up.  I'll check out your clean up.  Do you know what
happened?  I'm not super skilled with git so I want to understand what
hapened so I don't do that again.

-Andy S.

**Comment by aespielberg** [comment]
Hi Twan,

Some changes were made to dynamics which you removed which tried to
reconcile calling dynamics on the msspoly class (not valid) by returning a
sym instead of an msspoly.  Did you want to remove that?

-Andy S.

**Comment by aespielberg** [comment]
I'm going to create a new pull request so I don't obliterate my other
branch, although I want to make another push once I know what you'd like me
to do about the dynamics stuff.

-Andy S.

**Comment by tkoolen** [comment]
I did not mean to remove any functionality you tried to introduce, I just wanted to get rid of some bad merge commits by rebasing the following commits onto the master branch:
* c8a6601 (Committing progress on parameterizing kinematics)
* be6cd2e (committing any uncommitted changes)
* 4516a5d (Committing lots of changes to try to make dynamics work with msspoly. Fails near end due to inability to take inverse)
* 5898d19 (Fixed overaggressive msspoly errors)

This image shows the conceptual difference between rebasing and merging:
Rebase vs Merge
where C represents your commits.
I did this using an interactive rebase:
```
git checkout -b tk-rebase-aespielberg-msspoly_dynamics aespielberg/msspoly_dynamics
git rebase -i master
```
then I commented out everything except the commits above, saved the file, and exited my editor. There was one whitespace conflict during the rebase, which was easy to resolve.

A couple of questions regarding the missing dynamics change:
* what file/class were you referring to when you said "Some changes were made to dynamics [...]"? what is `dynamics` a method of?
* was any of the work you're trying to get into master committed by somebody other than yourself?

**Comment by aespielberg** [comment]
Ah okay, thank you for the info.  Next time I see weird errors like this, I
will see if I did a weird divergent thing and try rebasing.

dynamics is part of manipulator.m
All the work I'm trying to get into master should be committed by me.  I'm
the only one working on my branch.

-Andy S.

**Comment by aespielberg** [comment]
Should I add back my changes to dynamics then or would you like to look it
over?

-Andy S.

**Comment by tkoolen** [comment]
Please give me a minute to investigate a little more.

By the way, I think what maybe happened with your merges is that you added drake-distro as a remote for drake and then tried to merge drake-distro master into your drake branch somehow. Compare the files and folders in https://github.com/aespielberg/drake/tree/msspoly_dynamics to those in https://github.com/RobotLocomotion/drake/ and https://github.com/RobotLocomotion/drake-distro/. Your branch is kind of a Frankenstein mix of both.

Could you:

1. enter `git remote show` in /home/aespielberg/drake/drake-distro/drake/
1. for all remotes printed out by the previous command, type `git remote show <name of remote>` and post the results.

**Comment by aespielberg** [comment]
No prob!

aespielberg@ubuntu:~/drake/drake-distro/drake$ git remote show
origin
twan
upstream
aespielberg@ubuntu:~/drake/drake-distro/drake$ git remote show origin
* remote origin
  Fetch URL: git@github.com:aespielberg/drake.git
  Push  URL: git@github.com:aespielberg/drake.git
  HEAD branch: master
  Remote branches:
    master                                         tracked
    msspoly_dynamics                               tracked
    tk-rebase-aespielberg-msspoly_dynamics-cleaned tracked
  Local refs configured for 'git push':
    master                                         pushes to master
                                (up to date)
    msspoly_dynamics                               pushes to
msspoly_dynamics                               (up to date)
    tk-rebase-aespielberg-msspoly_dynamics-cleaned pushes to
tk-rebase-aespielberg-msspoly_dynamics-cleaned (up to date)
aespielberg@ubuntu:~/drake/drake-distro/drake$ git remote show twan
* remote twan
  Fetch URL: https://github.com/tkoolen/drake.git
  Push  URL: https://github.com/tkoolen/drake.git
  HEAD branch: master
  Remote branches:
    body-wrench-in-qp                         new (next fetch will store in
remotes/twan)
    cmake-project-fixes                       new (next fetch will store in
remotes/twan)
    cmpControl                                new (next fetch will store in
remotes/twan)
    drakeGradientUtil                         new (next fetch will store in
remotes/twan)
    drakeJoint                                new (next fetch will store in
remotes/twan)
    drakeJointInRigidBody                     new (next fetch will store in
remotes/twan)
    drakeJointInRigidBodyRebased              new (next fetch will store in
remotes/twan)
    fix-msvc-compile-warnings                 new (next fetch will store in
remotes/twan)
    fix-self-assignment                       new (next fetch will store in
remotes/twan)
    fix-testDrakeGradientUtil                 new (next fetch will store in
remotes/twan)
    fixDRCBuild                               new (next fetch will store in
remotes/twan)
    forwardKin-cleanup                        new (next fetch will store in
remotes/twan)
    geometricJacobian-cpp                     new (next fetch will store in
remotes/twan)
    geometricJacobianGradient                 new (next fetch will store in
remotes/twan)
    integrate-joint-position                  new (next fetch will store in
remotes/twan)
    issue-311                                 new (next fetch will store in
remotes/twan)
    issue-369                                 new (next fetch will store in
remotes/twan)
    issue-469                                 new (next fetch will store in
remotes/twan)
    issue-774                                 new (next fetch will store in
remotes/twan)
    issue-889                                 new (next fetch will store in
remotes/twan)
    joint-calibration-updates                 new (next fetch will store in
remotes/twan)
    joint-calibration-updates-2               new (next fetch will store in
remotes/twan)
    master                                    new (next fetch will store in
remotes/twan)
    new-message-monitors                      new (next fetch will store in
remotes/twan)
    newKinsol                                 new (next fetch will store in
remotes/twan)
    newKinsolForwardKin                       new (next fetch will store in
remotes/twan)
    old-kinsol-call-new-kinsol                new (next fetch will store in
remotes/twan)
    pack-mex-util                             new (next fetch will store in
remotes/twan)
    purge-old-kinematics                      new (next fetch will store in
remotes/twan)
    qpLocomotionPlan-cpp-rebase-unwrap-expmap new (next fetch will store in
remotes/twan)
    qpSpline-linear-system-mex                new (next fetch will store in
remotes/twan)
    quat-dot-times-v-issue-1002               new (next fetch will store in
remotes/twan)
    quat-dot-times-v-issue-1002-rebased       new (next fetch will store in
remotes/twan)
    rd-smooth-swing                           new (next fetch will store in
remotes/twan)
    resolveCenterOfPressure-updates           new (next fetch will store in
remotes/twan)
    smooth-swing-mexed-qpSpline               new (next fetch will store in
remotes/twan)
    task-space-control                        new (next fetch will store in
remotes/twan)
    tk-autodiff-scalar                        new (next fetch will store in
remotes/twan)
    tk-calibration-using-NLP-tweaks           new (next fetch will store in
remotes/twan)
    tk-drakeJoint-templated                   new (next fetch will store in
remotes/twan)
    tk-fix-win-rd-merge-drc                   new (next fetch will store in
remotes/twan)
    tk-ik-pelvis-traj                         new (next fetch will store in
remotes/twan)
    tk-purge-old-kinematics                   new (next fetch will store in
remotes/twan)
    tk-rebase-aespielberg-msspoly_dynamics    new (next fetch will store in
remotes/twan)
  Local refs configured for 'git push':
    master                                 pushes to master
                (fast-forwardable)
    tk-rebase-aespielberg-msspoly_dynamics pushes to
tk-rebase-aespielberg-msspoly_dynamics (fast-forwardable)
aespielberg@ubuntu:~/drake/drake-distro/drake$ git remote show upstream
* remote upstream
  Fetch URL: https://github.com/RobotLocomotion/drake.git
  Push  URL: https://github.com/RobotLocomotion/drake.git
  HEAD branch: master
  Remote branch:
    master tracked
  Local branch configured for 'git pull':
    master merges with remote master
  Local ref configured for 'git push':
    master pushes to master (local out of date)

**Comment by tkoolen** [comment]
Hmm strange, I was expecting the fetch and push URLs for upstream to point to drake-distro. I'm calling it a night, I'll look at it some more tomorrow.

**Comment by tkoolen** [comment]
Since you have the other pull request open now, could you close this one?

**Comment by aespielberg** [comment]
Oops, sorry, sure!

Andrew Spielberg
PhD Student
MIT - Computer Science and Artificial Intelligence Laboratory

**Pull request closed without merging**